## Table of Contents 

### 01 Importing and Checking the Data
### 02 Data Wrangling
### 03 Data Cleaning
- 3.1 Checking for mixed-type columns
- 3.2 Finding and addressing missing values
- 3.3 Finding and addressing duplicates

### 04 Checking and Exporting the Data

## 01 Importing and Checking the Data

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create a project folder string

path = r'/Users/nora/Desktop/CareerFoundry Achievement 6'

In [3]:
# Import the data set

df_airbnb = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'airbnb_original.csv'))

In [4]:
df_airbnb.head(100)

id                                              name  host_id  \
0    2539                Clean & quiet apt home by the park     2787   
1    2595                             Skylit Midtown Castle     2845   
2    3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3    3831                   Cozy Entire Floor of Brownstone     4869   
4    5022  Entire Apt: Spacious Studio/Loft by central park     7192   
..    ...                                               ...      ...   
95  21293                    Sunny Apartment in Artist Home    44145   
96  21456                Light-filled classic Central Park     42032   
97  21644                         Upper Manhattan, New York    82685   
98  21794                       COZY QUIET room 4 DOOGLERS!    83257   
99  22911               The Stuydio Modern and Light Filled    87773   

      host_name neighbourhood_group       neighbourhood  latitude  longitude  \
0          John            Brooklyn          Kensington  40.64749  -73.97237   
1      Jennifer           Manhattan             Midtown  40.75362  -73.98377   
2     Elisabeth           Manhattan              Harlem  40.80902  -73.94190   
3   LisaRoxanne            Brooklyn        Clinton Hill  40.68514  -73.95976   
4         Laura           Manhattan         East Harlem  40.79851  -73.94399   
..          ...                 ...                 ...       ...        ...   
95       Tyrome            Brooklyn            Bushwick  40.70093  -73.92609   
96         Dana           Manhattan     Upper West Side  40.79764  -73.96177   
97      Elliott           Manhattan              Harlem  40.82803  -73.94731   
98         Olan           Manhattan             Chelsea  40.74008  -74.00271   
99       Shelly            Brooklyn  Bedford-Stuyvesant  40.68413  -73.92357   

          room_type  price  minimum_nights  number_of_reviews last_review  \
0      Private room    149               1                  9  2018-10-19   
1   Entire home/apt    225               1                 45  2019-05-21   
2      Private room    150               3                  0         NaN   
3   Entire home/apt     89               1                270  2019-07-05   
4   Entire home/apt     80              10                  9  2018-11-19   
..              ...    ...             ...                ...         ...   
95  Entire home/apt    105               3                118  2019-06-18   
96  Entire home/apt    140               3                 81  2019-07-07   
97     Private room     89               1                  1  2018-10-09   
98     Private room     98              30                 30  2019-05-01   
99  Entire home/apt    125               7                139  2018-10-28   

    reviews_per_month  calculated_host_listings_count  availability_365  
0                0.21                               6               365  
1                0.38                               2               355  
2                 NaN                               1               365  
3                4.64                               1               194  
4                0.10                               1                 0  
..                ...                             ...               ...  
95               1.05                               2                 9  
96               0.71                               1               198  
97               0.11                               1               365  
98               0.27                               2               364  
99               1.23                               2               311  

[100 rows x 16 columns]

In [5]:
# Check dimensions of data set

df_airbnb.shape

(48895, 16)

In [6]:
# Summary statistics of uncleaned data set

df_summary_stats = df_airbnb.describe()

In [7]:
df_summary_stats

id       host_id      latitude     longitude         price  \
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000   
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687   
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000   
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000   
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48895.000000       48895.000000       38843.000000   
mean         7.029962          23.274466           1.373221   
std         20.510550          44.550582           1.680442   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

In [8]:
df_summary_stats.to_clipboard()

## 02 Data Wrangling 

In [9]:
# defining a new data frame for cleaned data

df_airbnb_clean = df_airbnb

In [11]:
# Get frequency counts of number_of_reviews column 

df_airbnb_clean['reviews_per_month'].value_counts().sort_index()

0.01      42
0.02     919
0.03     804
0.04     655
0.05     893
        ... 
17.82      1
19.75      1
20.94      1
27.95      1
58.50      1
Name: reviews_per_month, Length: 937, dtype: int64

58.50 is an unrealistic value for monthly reviews considering every guest stays at least 1 night. 

In [12]:
# Dropping row where reviews_per_month == 58.50 

df_airbnb_clean.drop(df_airbnb_clean.loc[df_airbnb_clean['reviews_per_month']==58.50].index, inplace=True)

In [13]:
# Get frequency counts of neighbourhood_group column 

df_airbnb_clean['neighbourhood_group'].value_counts()

Manhattan        21660
Brooklyn         20104
Queens            5666
Bronx             1091
Staten Island      373
Name: neighbourhood_group, dtype: int64

In [14]:
# Get frequency counts of neighbourhood column 

pd.set_option('display.max_rows', None)

df_airbnb_clean['neighbourhood'].value_counts().sort_index()

Allerton                        42
Arden Heights                    4
Arrochar                        21
Arverne                         77
Astoria                        900
Bath Beach                      17
Battery Park City               70
Bay Ridge                      141
Bay Terrace                      6
Bay Terrace, Staten Island       2
Baychester                       7
Bayside                         39
Bayswater                       17
Bedford-Stuyvesant            3714
Belle Harbor                     8
Bellerose                       14
Belmont                         24
Bensonhurst                     75
Bergen Beach                    10
Boerum Hill                    177
Borough Park                   136
Breezy Point                     3
Briarwood                       56
Brighton Beach                  75
Bronxdale                       19
Brooklyn Heights               154
Brownsville                     61
Bull's Head                      6
Bushwick            

In [15]:
# Get frequency counts of room_type column 

df_airbnb_clean['room_type'].value_counts(dropna=False)

Entire home/apt    25409
Private room       22325
Shared room         1160
Name: room_type, dtype: int64

## 03 Data Cleaning

### 3.1 Checking for mixed-type columns

In [16]:
#Check for mixed-type columns

for col in df_airbnb_clean.columns.tolist():
  weird = (df_airbnb_clean[[col]].applymap(type) != df_airbnb_clean[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_airbnb_clean[weird]) > 0:
    print (col)

name
host_name
last_review


In [17]:
# Fix the mixed-type columns

df_airbnb_clean['name'] = df_airbnb_clean['name'].astype('str')
df_airbnb_clean['host_name'] = df_airbnb_clean['host_name'].astype('str')

### 3.2 Finding and addressing missing values 

In [18]:
# Find missing values

df_airbnb_clean.isnull().sum()

id                                    0
name                                  0
host_id                               0
host_name                             0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

There are 10052 missing values in both the last_review and the reviews_per_month column. These values could be missing because there are new listings, which do not have any reviews yet. We could address this issue by creating a new column, which indicates whether a listing has any reviews.

In [19]:
# Defining new column 

df_airbnb_clean['no_reviews'] = df_airbnb_clean['last_review'].isnull() == True

In [20]:
# Checking the data frame 

df_airbnb_clean.head(100)

id                                               name  host_id  \
0    2539                 Clean & quiet apt home by the park     2787   
1    2595                              Skylit Midtown Castle     2845   
2    3647                THE VILLAGE OF HARLEM....NEW YORK !     4632   
3    3831                    Cozy Entire Floor of Brownstone     4869   
4    5022   Entire Apt: Spacious Studio/Loft by central park     7192   
5    5099          Large Cozy 1 BR Apartment In Midtown East     7322   
6    5121                                    BlissArtsSpace!     7356   
7    5178                   Large Furnished Room Near B'way      8967   
8    5203                 Cozy Clean Guest Room - Family Apt     7490   
9    5238                 Cute & Cozy Lower East Side 1 bdrm     7549   
10   5295                   Beautiful 1br on Upper West Side     7702   
11   5441                    Central Manhattan/near Broadway     7989   
12   5803     Lovely Room 1, Garden, Best Area, Legal rental     9744   
13   6021   Wonderful Guest Bedroom in Manhattan for SINGLES    11528   
14   6090                      West Village Nest - Superhost    11975   
15   6848                   Only 2 stops to Manhattan studio    15991   
16   7097                  Perfect for Your Parents + Garden    17571   
17   7322                                    Chelsea Perfect    18946   
18   7726    Hip Historic Brownstone Apartment with Backyard    20950   
19   7750                  Huge 2 BR Upper East  Cental Park    17985   
20   7801                   Sweet and Spacious Brooklyn Loft    21207   
21   8024                     CBG CtyBGd HelpsHaiti rm#1:1-4    22486   
22   8025                           CBG Helps Haiti Room#2.5    22486   
23   8110                              CBG Helps Haiti Rm #2    22486   
24   8490             MAISON DES SIRENES1,bohemian apartment    25183   
25   8505                 Sunny Bedroom Across Prospect Park    25326   
26   8700  Magnifique Suite au N de Manhattan - vue Cloitres    26394   
27   9357                               Midtown Pied-a-terre    30193   
28   9518       SPACIOUS, LOVELY FURNISHED MANHATTAN BEDROOM    31374   
29   9657                   Modern 1 BR / NYC / EAST VILLAGE    21904   
30   9668                              front room/double bed    32294   
31   9704                Spacious 1 bedroom in luxe building    32045   
32   9782                  Loft in Williamsburg Area w/ Roof    32169   
33   9783                                back room/bunk beds    32294   
34  10452                              Large B&B Style rooms    35935   
35  10962    Lovely room 2 & garden; Best area, Legal rental     9744   
36  11452                        Clean and Quiet in Brooklyn     7355   
37  11708                          Cute apt in artist's home    44145   
38  11943                          Country space in the city    45445   
39  12048                LowerEastSide apt share shortterm 1     7549   
40  12192                                ENJOY Downtown NYC!    46978   
41  12299                Beautiful Sunny Park Slope Brooklyn    47610   
42  12303                  1bdr w private bath. in lofty apt    47618   
43  12318                                  West Side Retreat    16800   
44  12343                                 BEST BET IN HARLEM    47727   
45  12627  Entire apartment in central Brooklyn neighborh...    49670   
46  12937  1 Stop fr. Manhattan! Private Suite,Landmark B...    50124   
47  12940                 Charming Brownstone 3 - Near PRATT    50148   
48  13050                          bright and stylish duplex    50846   
49  13394                            Fort Greene brownstone     52335   
50  13808    Blue Room for 2 in Brownstone for $1350 monthly    54275   
51  14287     Cozy 1BD on Central Park West in New York City    56094   
52  14290                         * ORIGINAL BROOKLYN LOFT *    56104   
53  14314                    Greenpoint Place...Has It All!     5

### 3.3 Finding and addressing duplicates

In [21]:
# Check for duplicates 

df_dups = df_airbnb[df_airbnb.duplicated()]

In [22]:
df_dups.shape

(0, 17)

There are no duplicates in this data set.

## 04 Checking and Exporting the Data

In [23]:
# Check dimensions of data set

df_airbnb_clean.shape

(48894, 17)

In [24]:
# Summary statistics of cleaned data set

df_airbnb_clean.describe()

id       host_id      latitude     longitude         price  \
count  4.889400e+04  4.889400e+04  48894.000000  48894.000000  48894.000000   
mean   1.901686e+07  6.761640e+07     40.728948    -73.952169    152.721765   
std    1.098305e+07  7.860771e+07      0.054530      0.046157    240.156507   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471919e+06  7.821708e+06     40.690100    -73.983070     69.000000   
50%    1.967725e+07  3.079257e+07     40.723070    -73.955680    106.000000   
75%    2.915194e+07  1.074344e+08     40.763117    -73.936273    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48894.000000       48894.000000       38842.000000   
mean         7.030085          23.271751           1.371751   
std         20.510741          44.546994           1.655275   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          27.950000   

       calculated_host_listings_count  availability_365  
count                    48894.000000      48894.000000  
mean                         7.143944        112.777519  
std                         32.952855        131.620941  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

In [33]:
# Export cleaned data set

df_airbnb_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'airbnb_clean.csv'))